scc/light_client/provider/rpc_test.go

```diff
@@ -0,0 +1,318 @@
+package provider
+
+import (
+	"fmt"
+	"testing"
+
+	"github.com/0xsoniclabs/sonic/ethapi"
+	"github.com/0xsoniclabs/sonic/scc"
+	"github.com/0xsoniclabs/sonic/scc/bls"
+	"github.com/0xsoniclabs/sonic/scc/cert"
+	"github.com/Fantom-foundation/lachesis-base/inter/idx"
+	"github.com/golang/mock/gomock"
+	"github.com/stretchr/testify/require"
+)
+
+func TestRpcProvider_CanInitializeFromUrl(t *testing.T) {
+	require := require.New(t)
+
+	provider, err := NewRpcProviderFromURL("http://localhost:8545")
+	t.Cleanup(provider.Close)
+	require.NoError(err)
+	require.NotNil(provider)
+	require.False(provider.IsClosed())
+}
+
+func TestRpcProvider_ReportsErrorForNilClient(t *testing.T) {
+	require := require.New(t)
+
+	provider := NewRpcProviderFromClient(nil)
+
+	// get committee certificates
+	_, err := provider.GetCommitteeCertificates(0, 1)
+	require.Error(err)
+
+	// get block certificates
+	_, err = provider.GetBlockCertificates(0, 1)
+	require.Error(err)
+}
+
+func TestRpcProvider_IsClosed_Reports(t *testing.T) {
+	require := require.New(t)
+	ctrl := gomock.NewController(t)
+
+	t.Run("provider with nil client is closed", func(t *testing.T) {
+		provider := NewRpcProviderFromClient(nil)
+		require.True(provider.IsClosed())
+	})
+
+	t.Run("provider with client is not closed", func(t *testing.T) {
+		client := NewMockRpcClient(ctrl)
+		client.EXPECT().Close().AnyTimes()
+		provider := NewRpcProviderFromClient(client)
+		require.False(provider.IsClosed())
+		provider.Close()
+	})
+
+	t.Run("provider with client can be closed", func(t *testing.T) {
+		client := NewMockRpcClient(ctrl)
+		client.EXPECT().Close()
+		provider := NewRpcProviderFromClient(client)
+		require.False(provider.IsClosed())
+		provider.Close()
+		require.True(provider.IsClosed())
+	})
+
+	t.Run("closed provider can be re-closed", func(t *testing.T) {
+		client := NewMockRpcClient(ctrl)
+		client.EXPECT().Close().AnyTimes()
+		provider := NewRpcProviderFromClient(client)
+		provider.Close()
+		require.True(provider.IsClosed())
+		provider.Close()
+		require.True(provider.IsClosed())
+	})
+}
+
+func TestRpcProvider_FailsToRequestAfterClose(t *testing.T) {
+	require := require.New(t)
+	ctrl := gomock.NewController(t)
+	client := NewMockRpcClient(ctrl)
+
+	provider := NewRpcProviderFromClient(client)
+
+	// close provider
+	client.EXPECT().Close()
+	provider.Close()
+
+	// get committee certificates
+	_, err := provider.GetCommitteeCertificates(0, 1)
+	require.Error(err)
+
+	// get block certificates
+	_, err = provider.GetBlockCertificates(0, 1)
+	require.Error(err)
+}
+
+func TestRpcProvider_NewRpcProvider_ReportsErrorForInvalidURL(t *testing.T) {
```

Comment:
small nit: keep tests for the same functions grouped if possible

Reply:
https://github.com/0xsoniclabs/sonic/commit/fef0b75b24cc8a2a8e85fa04beff49bfb96cd32f